src/main.cpp

```diff
@@ -40,270 +26,264 @@ void setup(){
 
   LCD.begin(16,2) ;
 
-  Serial.begin(9600) ;
+  Serial.begin(115200) ;
 
   RCServo0.attach(SERVO_0());
   RCServo1.attach(SERVO_1());
   RCServo2.attach(SERVO_2());
   RCServo3.attach(SERVO_3());
-
-  left_surface = false;  // TODO read the switch
+  
+  left_surface = true;  // TODO read the switch
   // left_surface = digitalRead(LEFT_RIGHT_SWITCH());
+
+  hardware::Qrd qrd;
+  hardware::Beacon beacon;
+  hardware::Encoder l_encoder(0);
+  hardware::Encoder r_encoder(1, true);
+
+  hardware::EncoderMotor l_motor(L_MOTOR(), l_encoder, true);
+  hardware::EncoderMotor r_motor(R_MOTOR(), r_encoder);
+
+  sequences::Claw claw;
+  sequences::Drivetrain driver(l_motor, r_motor, qrd);
+  sequences::Platform platform(driver);
+
+  stages::Gate gate(driver, beacon, r_encoder); // TODO change encoder dynamically
+  stages::Pickup pickup(qrd, r_encoder, claw, driver);
+  stages::Zipline zipline(driver, platform, beacon, r_encoder, qrd);
+
   claw.fold(left_surface);
   delay(1000);
 
-  // claw.raise(left_surface ? sequences::RIGHT_CLAW : sequences::LEFT_CLAW);
-  // while(!claw.loop()) delay(100);
-  // claw.raise(left_surface ? sequences::LEFT_CLAW : sequences::RIGHT_CLAW);
-  // while(!claw.loop()) delay(100);
+  unsigned long waitUntil = millis() + loop_delay;
+  
+  enableEncoderInterrupts();
+
+  ////////////////////////////////
+  // Insert extra variable here //
+  ////////////////////////////////
+  uint8_t state = 0;
+
+  ////////////////////////////////
+  for (;;) {//////////////////////
+  ////////////////////////////////
+  //// Insert code here //////////
+  ////////////////////////////////
+
+    switch (state) {
+      case 0:  // wait for start button
+        if (!startbutton()) break;
+        else state++;
+      case 1:  // going through gate
+        if (gate.loop())
+          state++;
+        break;
+      case 2:  // picking up agents
+        if (pickup.loop())
+          state++;
+        break;
+      case 3:  // finding zipline
+         if (zipline.loop())
+           state++;
+         break;
+    }
+  ///////////////////////////////
+  //// Do not touch /////////////
+  ///////////////////////////////
+    while (millis() < waitUntil) continue;
+    waitUntil += loop_delay;
```

Comment:
safer to do `waitUntil = millis() + loop_delay` just in case we have some blocking code somewhere?